source/extensions/filters/http/admission_control/thread_local_controller.cc

```diff
@@ -0,0 +1,49 @@
+#include "extensions/filters/http/admission_control/thread_local_controller.h"
+
+#include <cstdint>
+
+#include "envoy/common/pure.h"
+#include "envoy/common/time.h"
+#include "envoy/http/codes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+
+static constexpr std::chrono::seconds defaultHistoryGranularity{1};
+
+ThreadLocalControllerImpl::ThreadLocalControllerImpl(TimeSource& time_source,
+                                                     std::chrono::seconds sampling_window)
+    : time_source_(time_source), sampling_window_(sampling_window) {}
+
+void ThreadLocalControllerImpl::maybeUpdateHistoricalData() {
+  // Purge stale samples.
+  while (!historical_data_.empty() && ageOfOldestSample() >= sampling_window_) {
+    removeOldestSample();
+  }
+
+  // It's possible we purged stale samples from the history and are left with nothing, so it's
+  // necessary to add an empty entry. We will also need to roll over into a new entry in the
+  // historical data if we've exceeded the time specified by the granularity.
```

Comment:
its not obvious to me how this granularity comes into play, maybe add some detail?